crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3761,22 +3762,22 @@ impl<'db> TypeInferenceBuilder<'db> {
 impl<'db> TypeInferenceBuilder<'db> {
     fn infer_annotation_expression(&mut self, expression: &ast::Expr) -> Type<'db> {
         // https://typing.readthedocs.io/en/latest/spec/annotations.html#grammar-token-expression-grammar-annotation_expression
-        match expression {
+        let annotation_ty = match expression {
             // TODO: parse the expression and check whether it is a string annotation, since they
             // can be annotation expressions distinct from type expressions.
             // https://typing.readthedocs.io/en/latest/spec/annotations.html#string-annotations
-            ast::Expr::StringLiteral(_literal) => {
-                self.store_expression_type(expression, Type::Todo);
-                Type::Todo
-            }
+            ast::Expr::StringLiteral(_literal) => Type::Todo,
 
             // Annotation expressions also get special handling for `*args` and `**kwargs`.
             ast::Expr::Starred(starred) => self.infer_starred_expression(starred),
 
             // All other annotation expressions are (possibly) valid type expressions, so handle
             // them there instead.
             type_expr => self.infer_type_expression(type_expr),
-        }
+        };
+
+        self.store_expression_type(expression, annotation_ty);
+        annotation_ty
```

Comment:
This is probably the easiest solution to the *"we don't want to risk forgetting to store the annotation type in any of the non-`infer_type_expression` branches above"* problem that I could think of. The way this works is by having a new function *infer_**and_store**_type_expression*, while the previous *infer_type_expression* only returns the type. If you feel that this introduces other risks (calling the wrong function elsewhere), let me know.

Another solution (also not very elegant) would be to explicitly state whether or not the expression type should be stored, as a kind of `mode` parameter to the `infer_type_expression` function (which would either be `InferOnly` or `InferAndStore`, or similar).

A third option would be to simply ignore double-insertions.

Reply:
> A third option would be to simply ignore double-insertions.

We could disable this in release build but the assertion is intentional to catch double-inference early.

Reply:
I like the option you've chosen here, but I think we should flip the naming, so that `infer_type_expression` and `infer_expression` (the "default" options someone is likely to reach for) _do_ store the expression type.

I realize the current naming you have here is sensible because storing is doing an extra thing, so it makes sense for that to be reflected in the name. But I think the need that `infer_annotation_expression` has to infer a type expression without storing it is the unusual case, and it's important that the method we should almost always prefer be the one with the shorter/simpler name.

So I would suggest that `infer_expression_type` does store, and we have `infer_expression_type_impl` or `infer_expression_type_no_store` (or perhaps you have a better idea for the awkward naming here) for the unusual case.

Reply:
> So I would suggest that `infer_expression_type` does store, and we have `infer_expression_type_impl` or `infer_expression_type_no_store`

Done